File: services/api/app/services/names.py
"""
Student-name normalization.

Plays store ``learner_label`` as a plain string and the roll gradebook
groups plays by exact match against the current roster, so an invisible
whitespace difference (e.g. a double space pasted into the roster)
silently orphans a student's plays.  Every comparison between a
learner-supplied name and a roster name — and every stored
``learner_label`` — must go through :func:`normalize_student_name`.

Roster names themselves are stored as the teacher typed them; they are
normalized at comparison time only, so existing rolls keep working
without a data migration.
"""

from __future__ import annotations


def normalize_student_name(value: str | None) -> str:
    """Collapse internal whitespace runs to single spaces and trim.

    Returns ``""`` for ``None`` or whitespace-only input.
    """
    if not value:
        return ""
    return " ".join(value.split())


def student_name_key(value: str | None) -> str:
    """Reduce a student name to a cross-app comparison key.

    Casefolds, collapses whitespace, and folds ``"Last, First"`` to
    ``"first last"`` so the same student matches across apps whose rosters
    disagree only on name order or case.  Shared by copy with the
    essay-grader repo — keep the two implementations identical.

    Limitation: a comma-suffixed name ("Smith, Jr., John") folds wrong;
    acceptable under the identical-roster-spellings convention.
    """
    s = normalize_student_name(value).casefold()
    if "," in s:
        last, _, first = s.partition(",")
        s = f"{first.strip()} {last.strip()}"
    return " ".join(s.split())


def names_equivalent(a: str | None, b: str | None) -> bool:
    """True when two spellings identify the same student (never for blanks)."""
    ka, kb = student_name_key(a), student_name_key(b)
    return bool(ka) and ka == kb

File: apps/web/lib/names.ts
/**
 * Student-name comparison helpers.
 *
 * TypeScript port of services/api/app/services/names.py — keep the two in
 * sync (and the copies in the essay-grader repo). Casefolds, collapses
 * whitespace, and folds "Last, First" to "first last" so the same student
 * matches across apps whose rosters disagree only on order or case.
 *
 * Limitation: comma-suffixed names ("Smith, Jr., John") fold wrong;
 * acceptable under the identical-roster-spellings convention.
 */

export function normalizeStudentName(value: string | null | undefined): string {
  if (!value) return "";
  return value.split(/\s+/).filter(Boolean).join(" ");
}

export function studentNameKey(value: string | null | undefined): string {
  let s = normalizeStudentName(value).toLowerCase();
  const comma = s.indexOf(",");
  if (comma !== -1) {
    const last = s.slice(0, comma).trim();
    const first = s.slice(comma + 1).trim();
    s = `${first} ${last}`;
  }
  return s.split(/\s+/).filter(Boolean).join(" ");
}

export function namesEquivalent(
  a: string | null | undefined,
  b: string | null | undefined,
): boolean {
  const ka = studentNameKey(a);
  return ka.length > 0 && ka === studentNameKey(b);
}
